### PR TITLE
multiple route tables support

### LINKED ISSUE #2637: multiple route tables support
### Feature request

Shell we have multiple route tables in VPC level and can be associate/disassociate with Subnet.
I think this features is already in OVN (set static route with --route_table in lr and set option in lrp)

### Use case

- reuse routings and use different route table in each Subnet
- associate / disassociate routing table in Subnet easily

I think use policy-based route with src=<subnet_cidr> match can instead, but will be more flow entry and difficulty to switch route table

I can contribute to this feature If that make sense

### PR DESCRIPTION
**Pull request by mingoooo**
- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

### What type of this PR
Examples of user facing changes:
- Features

### Which issue(s) this PR fixes:
Fixes #2637

### <samp>🤖 Generated by Copilot at 8db39e3</samp>

This pull request adds support for custom route tables in VPCs and subnets. It defines a new `RouteTable` API type and modifies the `VpcSpec` and `SubnetStatus` types to reference it. It also updates the generated code for these types.

### <samp>🤖 Generated by Copilot at 8db39e3</samp>

> _`VpcSpec` and `RouteTable`_
> _New types for custom routes_
> _Autumn leaves `DeepCopyInto`_

### <samp>🤖 Generated by Copilot at 8db39e3</samp>

*  Add support for route tables in subnets and VPCs
  * Add `RouteTable` field to `SubnetStatus` type ([link](https://github.com/kubeovn/kube-ovn/pull/2684/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aR143-R144))
  * Add `RouteTables` field to `VpcSpec` type ([link](https://github.com/kubeovn/kube-ovn/pull/2684/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aR362))
  * Define `RouteTable` type with `Name` and `Routes` fields ([link](https://github.com/kubeovn/kube-ovn/pull/2684/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aR385-R389))
  * Generate `DeepCopyInto` function for `RouteTable` type ([link](https://github.com/kubeovn/kube-ovn/pull/2684/files?diff=unified&w=0#diff-06f8e5bc84406eb8a89c72c345fc4ff5f144e480103ad55d681f7567f0f85515R1399-R1425))
  * Copy `RouteTables` field in `VpcSpec` `DeepCopyInto` function ([link](https://github.com/kubeovn/kube-ovn/pull/2684/files?diff=unified&w=0#diff-06f8e5bc84406eb8a89c72c345fc4ff5f144e480103ad55d681f7567f0f85515R2453-R2463))

### CODE REVIEW

File: pkg/controller/subnet.go [hunk 1]
```diff
@@ -2511,3 +2538,32 @@ func (c *Controller) deletePolicyRouteForU2OInterconn(subnet *kubeovnv1.Subnet)
 
 	return nil
 }
+
+func (c *Controller) reconcileRouteTableForSubnet(subnet *kubeovnv1.Subnet) error {
+	klog.Infof("reconcile route table %s for subnet %s", subnet.Spec.RouteTable, subnet.Name)
```
Comment (zhangzujian) [inline pkg/controller/subnet.go:2543]:
```suggestion
	if subnet.Spec.Vlan != "" && !subnet.Spec.U2OInterconnection {
		return nil
	}

	klog.Infof("reconcile route table %q for subnet %s", subnet.Spec.RouteTable, subnet.Name)
```

**Review by bobz965 — COMMENTED** [review]
there are so many codes which use the static routes, and as you mentioned in the feature, `static route with --route_table` , In my view, the route table is a function in static routes,  I think extend or change vpc spec static routes is a better way, not just replaced the spec static routes.

**Comment by mingoooo** [comment]
> there are so many codes which use the static routes, and as you mentioned in the feature, `static route with --route_table` , In my view, the route table is a function in static routes, I think extend or change vpc spec static routes is a better way, not just replaced the spec static routes.

as mentioned in #2637, actually I'm tring add routeTable as option in staticRoute at the beginning of the design to avoid many codes changes, but found that - 

1. actually this is not a right staucture, user may difficult to view his route entries in one of the route table
2. when we need to add default or system route entry in each route table, or base on the table level to do something, that need to loop all of the static routes to do that, It doesn't seem very reasonable

**Comment by mingoooo** [comment]
@bobz965 hope that make sense, please correct me if I have any misunderstanding
@oilbeater @zhangzujian any comment on the crd design?

**Comment by bobz965** [comment]
> > there are so many codes which use the static routes, and as you mentioned in the feature, `static route with --route_table` , In my view, the route table is a function in static routes, I think extend or change vpc spec static routes is a better way, not just replaced the spec static routes.
> 
> as mentioned in #2637, actually I'm tring add routeTable as option in staticRoute at the beginning of the design to avoid many codes changes, but found that -
> 
> 1. actually this is not a right staucture, user may difficult to view his route entries in one of the route table
> 2. when we need to add default or system route entry in each route table, or base on the table level to do something, that need to loop all of the static routes to do that, It doesn't seem very reasonable

how about add the route table name into StaicRoute, after all, it is a option in a static route. 
image

**Comment by mingoooo** [comment]
> > > there are so many codes which use the static routes, and as you mentioned in the feature, `static route with --route_table` , In my view, the route table is a function in static routes, I think extend or change vpc spec static routes is a better way, not just replaced the spec static routes.
> > 
> > 
> > as mentioned in #2637, actually I'm tring add routeTable as option in staticRoute at the beginning of the design to avoid many codes changes, but found that -
> > 
> > 1. actually this is not a right staucture, user may difficult to view his route entries in one of the route table
> > 2. when we need to add default or system route entry in each route table, or base on the table level to do something, that need to loop all of the static routes to do that, It doesn't seem very reasonable
> 
> how about add the route table name into StaicRoute, after all, it is a option in a static route. image

get it, I'm just saying that it's a little cryptic and hard to understand for users to associate tables in this way.
It is difficult for the user to know which routes are included in the table, don't you think?
should we have an overview for table level like `ovn-ctl lr-route-list`

**Comment by bobz965** [comment]
> > > > there are so many codes which use the static routes, and as you mentioned in the feature, `static route with --route_table` , In my view, the route table is a function in static routes, I think extend or change vpc spec static routes is a better way, not just replaced the spec static routes.
> > > 
> > > 
> > > as mentioned in #2637, actually I'm tring add routeTable as option in staticRoute at the beginning of the design to avoid many codes changes, but found that -
> > > 
> > > 1. actually this is not a right staucture, user may difficult to view his route entries in one of the route table
> > > 2. when we need to add default or system route entry in each route table, or base on the table level to do something, that need to loop all of the static routes to do that, It doesn't seem very reasonable
> > 
> > 
> > how about add the route table name into StaicRoute, after all, it is a option in a static route. image
> 
> get it, I'm just saying that it's a little cryptic and hard to understand for users to associate tables in this way. It is difficult for the user to know which routes are included in the table, don't you think? should we have an overview for table level like `nbctl lr-route-list`

please attach the lr-route-list about the route table case， i still have no env with this feature

**Comment by mingoooo** [comment]
> > > > > there are so many codes which use the static routes, and as you mentioned in the feature, `static route with --route_table` , In my view, the route table is a function in static routes, I think extend or change vpc spec static routes is a better way, not just replaced the spec static routes.
> > > > 
> > > > 
> > > > as mentioned in #2637, actually I'm tring add routeTable as option in staticRoute at the beginning of the design to avoid many codes changes, but found that -
> > > > 
> > > > 1. actually this is not a right staucture, user may difficult to view his route entries in one of the route table
> > > > 2. when we need to add default or system route entry in each route table, or base on the table level to do something, that need to loop all of the static routes to do that, It doesn't seem very reasonable
> > > 
> > > 
> > > how about add the route table name into StaicRoute, after all, it is a option in a static route. image
> > 
> > 
> > get it, I'm just saying that it's a little cryptic and hard to understand for users to associate tables in this way. It is difficult for the user to know which routes are included in the table, don't you think? should we have an overview for table level like `nbctl lr-route-list`
> 
> please attach the lr-route-list about the route table case， i still have no env with this feature

```
k ko nbctl lr-route-list mingo-test
IPv4 Routes
Route Table <main>:
                  1.1.1.1                   1.1.1.1 dst-ip ecmp
                  1.1.1.1                   1.1.1.2 dst-ip ecmp
                  1.1.1.1                   1.1.1.3 dst-ip ecmp

Route Table mingo-rtb:
                  1.1.1.1                   2.2.2.2 dst-ip

Route Table rtb1:
                 1.1.1.11              192.168.11.1 dst-ip
```

you just need to add static route with `--route_table` option, than can be view multiple tables in `lr-route-list`

**Comment by mingoooo** [comment]
@bobz965 create a pr just what you want, please take a look #2748

**Comment by bobz965** [comment]
> route_table

ok

**Comment by bobz965** [comment]
> @bobz965 create a pr just what you want, please take a look #2748

ok

**Comment by mingoooo** [comment]
Duplicate of #2748

**Pull request closed without merging**